### DISCUSSION TITLE
LDM Notes for September 4th, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8417] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-09-04.md

### Agenda

- Triage (no milestone)
    - Type inference using method group natural type
    - Collection Expressions Next (C#13 and beyond)
    - Dictionary expressions
    - Extending patterns to "as"
    - `params in` parameters
- Triage (working set)
    - Only Allow Lexical Keywords in the Language
    - Permit variable declarations under disjunctive patterns
    - CallerCharacterNumberAttribute
    - `Task<T>` nullability covariance
    - Nullable analysis of LINQ queries

**Comment by MitchRazga** [comment]
Really liked this meetings notes. Was a good insight into the LDTs focus areas.

How does the LDM approach topics that are not yet proposals but are more about improving/expanding existing features?
Without a proposal, feedback can get spread over multiple discussions/issue comments which may be difficult to find when it comes time to look at that area again.

For example, something like ```ref record struct``` has been mentioned in a few different areas and is even mentioned in the [Type Unions Proposal](https://github.com/dotnet/csharplang/blob/7bfaafb048c89e6a8998938203e3ef24c8497cdc/proposals/TypeUnions.md?plain=1#L386) but there is no centralized discussion area since [it has no proposal](https://github.com/dotnet/csharplang/issues/7608#issuecomment-2079912688).

Should community members be [necroposting](https://en.wiktionary.org/wiki/necropost#English) on already implemented proposals to keep this feedback centralized? eg. https://github.com/dotnet/csharplang/issues/39
With GitHub Discussions, my concern is that since multiple threads may discuss the same topic but with a different title/keywords or if they have to many comments (as mentioned in the [readme ](https://github.com/dotnet/csharplang/blob/7bfaafb048c89e6a8998938203e3ef24c8497cdc/README.md?plain=1#L26)) they might be missed by the LDT.

Also, @333fred the link has broken after commit cd4d9a4a7cb17bcb36bf67ad5501296fd0d981d3

**Reply by 333fred** [reply]
Thanks for the ping, I meant to go update this yesterday but forgot.

As to things that are not yet proposals, it really varies on the member. Your best bet is to comment on some existing discussion, or opening a new one; commenting on an existing, implemented feature is likely just going to get comments that this feature has been implemented, please go open a new discussion. Several members of the LDM (myself included) subscribe to this repo and see every comment made on every discussion; some of us may not chime in very often, but we are watching.

**Reply by CyrusNajmabadi** [reply]
> but we are watching.

We're always watching wazowski.

**Comment by MgSam** [comment]
I'm confused about the feedback on https://github.com/dotnet/csharplang/issues/8210. The LDM feedback talks about how it would change `is`. Unless I'm missing something the proposal is entirely about expanding `as`, not changing `is`. 

Also, I think many people preferred the `when` syntax, which I think is highly consistent with the existing usages of `when` in C#. Should that be broken out into a separate proposal?

**Reply by 333fred** [reply]
> The LDM feedback talks about how it would change `is`.

Correct. That's exactly what it would do: collapse the true/false branches of a proper pattern match into `not null` and `null`. This then needs to be tested _again_ for `null`, rather than just using the original outcome of the `is`.

> Also, I think many people preferred the when syntax, which I think is highly consistent with the existing usages of when in C#. Should that be broken out into a separate proposal?

No. Both suffer the exact same problem. It is not the use of the word `as` that we dislike: it's the collapsing of condition.

**Reply by chrisoverzero** [reply]
The proposal is for the keyword spelled “ay-ess,” as is @MgSam’s question. But the responses both there and here are about the keyword spelled “eye-ess.”

Maybe the LDT can see three steps ahead where I cannot, but I don’t understand how the one got to the other.

**Reply by 333fred** [reply]
Sarcasm is not helpful. I'll say again, an `as pattern` is just `is pattern x ? x : null`, the result of which then immediately has to be checked for null. We do not like this.

**Reply by chrisoverzero** [reply]
That wasn’t sarcasm at all – I genuinely didn’t understand. Now I see the middle step, thank you. 

> I’ll say again […]

At the risk of upsetting you further, part of the reason I didn’t understand is that this was left out of the explanation. It’s a proposal about changing `as`, and the LDT’s response is about `is`. In your response, you confirmed that it was about a desire not to change `is`. I don’t think it’s unreasonable to wonder, since `is` wasn’t changing, whether some steps were skipped in the explanation because of the LDT’s expertise (alternatively, mis-reading), and it turns out that it seems like they were.

**Reply by 333fred** [reply]
I'm certainly willing to hear feedback on how to make the notes clearer. However, I don't know how read a comment like `The proposal is for the keyword spelled “ay-ess,”` except as sarcasm, as it implies that I don't know how to spell.

**Comment by jnm2** [comment]
A use case came up where I would have written:

```cs
var match = regex.Match(input) as { Success: True }
    ?? throw new InvalidOperationException(...);
```

**Reply by HaloFour** [reply]
Is that appreciably better than the following?

```cs
var match = regex.Match(input) switch {
        { Success: true } m => m,
        _ => throw new InvalidOperationException("boo")
};
```

**Reply by jnm2** [reply]
To me, yes. Two lines versus five. I know it could be four, but I'm not ready to generally start switching to K&R. Plus, `m`. Plus, for sure, more to visually parse.

**Reply by HaloFour** [reply]
Sure, it's less verbose and doesn't require an additional identifier, but is that "appreciably" better?  That is very subjective, I know.  Question is whether or not that pattern comes up often enough to justify its own feature.  I wouldn't mind a guard feature, like the ones proposed before pattern matching went leaky, which required an `else` clause.  I guess an `as` feature for matching that returns `null` is a step too far, though.

**Reply by jnm2** [reply]
It's the same amount of better as any existing usage of `as` with patterns that happen to just be types, especially paired with coalescing.

**Reply by HaloFour** [reply]
Honestly I kind of agree, `as`+`??` seems to cover those guard cases nicely, even given the concern of the `null` intermediate value.  I don't think the language can start requiring that they get paired up at this point, so we either accept the `null`, design a new language feature for this purpose, consider ways to "trim down" the use of `switch` or just live with the current state of things.

**Reply by jnm2** [reply]
I wouldn't see any value in requiring that they be paired, limiting composability. For instance, maybe you want to add on `?.Xyz` before coalescing with null. The language lets you combine features in ways that lead to poor understandability in a myriad ways already. Limiting the language's power to try to prevent people from writing confusing code seems like a dubious tradeoff with little prospect of big-picture success toward such a tenet.

**Reply by jnm2** [reply]
I'm currently at: I'd use it, I'd like using it, but so far it feels infrequent enough that I wouldn't be strongly in favor of investing in it, particularly over other things. For instance, I'd rather have compound assignments in initializers and `with`, and null-conditional assignments.

**Reply by CyrusNajmabadi** [reply]
> I'm currently at: I'd use it, I'd like using it, but so far it feels infrequent enough that I wouldn't be strongly in favor of investing in it, 

I'm in a similar boat.  And I personally think I'd use it so rarely that I think it's too cute and too niche to be worthwhile

**Reply by RikkiGibson** [reply]
In the world without `as`-patterns I think the following is the best way to write it

```cs
var match = regex.Match(input) is { Success: True } m ? m
    : throw new InvalidOperationException(...);
```

Or, in the specific case of wanting to assign to a local:

```cs
if (regex.Match(input) is not { Success: True } match)
    throw new InvalidOperationException(...);
```

**Reply by jnm2** [reply]
Seeing `var match =` at the start of the line has felt like an improvement to me, but it is such a special case. Not every guard can coalesce to `?? throw`.
